### PR TITLE
Add parent queue priority logic

### PR DESCRIPTION
**Pull request by Devidxyz**
## Current concept
Currently the **priority logic is only used in the child queues**: we prioritize the access flow child jobs over the status-update flow child jobs in a common access-check parent queue. The child job params are defined in the previous queues' result, so the priorities are also determined there. 
imageThe related code lines:
https://github.com/guildxyz/guild-backend/blob/main/src/flows/statusUpdate/workerFunctions/bulkPrepareManageReward.ts#L55
https://github.com/guildxyz/guild-backend/blob/main/src/flows/statusUpdate/workerFunctions/statusUpdatePreparation.ts#L131
(This was the easiest and most sensible solution at the time.)

## The problem
There's only 1 priority on the access-check parent queue, so there's only one priority list. If we start a bunch of status update jobs at a time, all the workers working on the access-check parent queue will be busy with the running status-update jobs and the "real-time" join jobs will have to wait for one (or more) status update job to finish.

## The solution
We don't set the child priorities in the previous queues. We add a new property to the Queue class, which determines what will be the next queue's priority, just like the properties `nextQueueName` and `nextQueueMap`, which did the same thing but with the next queue's name.
So if we had
`[queue1]--->[queue2]--->[queue3]  // priority is always 1`
now we can have something like 
`[queue1]---(+1 prio)---> [queue2]---(-1 prio)--->[queue3]`
image
This way we will have two different priorities in the access-check queue as we would expect normally, there's no unexpected tinkering with the child's priorities.

## Other info
- renamed `nextQueueMap` to `nextQueueNameMap` to avoid confusion
- companion core PR: https://github.com/guildxyz/guild-backend/pull/1204

### CODE REVIEW

File: src/flows/statusUpdate/getStatusUpdateFlowData.ts [hunk 1]
```diff
@@ -26,6 +26,7 @@ const getStatusUpdateFlowProps = (): FlowProps => {
         },
       ],
       nextQueueName: "access-check" as StatusUpdateFlowJob["queueName"],
+      nextQueuePriorityDiffMap: new Map([["status-update", +1]]), // increase priority by one
```
Comment (TomiOhl) [inline src/flows/statusUpdate/getStatusUpdateFlowData.ts]: Increasing priority equals lower priority, this comment made me think a bit now :D

File: src/flows/sharedQueues.ts [hunk 2]
```diff
@@ -9,10 +9,12 @@ export const manageRewardQueue = new Queue({
     "children:manage-reward:params",
     "children:manage-reward:jobs",
   ],
-  nextQueueMap: new Map([
+  priorities: 2,
```
Comment (TomiOhl) [inline src/flows/sharedQueues.ts:12]: priorities === priorities of child queues?

File: src/base/types.ts [hunk 3]
```diff
@@ -180,7 +180,8 @@ export type QueueOptions<NextQueueName extends string = string> = {
    * If the queue is part of multiple flows, it probably has multiple next queues
    * This map maps the flow names to the next queues
    */
-  nextQueueMap?: Map<string, string>;
+  nextQueueNameMap?: Map<string, string>;
+  nextQueuePriorityDiffMap?: Map<string, number>;
```
Comment (TomiOhl) [inline src/base/types.ts:184]: Consider adding an explanation comment here, just like for other properties

**Review by TomiOhl — APPROVED** [review]
Left three comments but makes sense and looks good to me

**Review by PAStheLoD — APPROVED** [review]
lgtm after some discord discussion :)

current bottleneck is at the `access-check` parent queue. (queues are simple FIFO, we use 2 simple ones to achieve prioritization.) this PR "enables" 2-level priority handling for a bunch of queues.

File: src/flows/sharedQueues.ts (hunk 2 above)
Comment (Devidxyz) [reply src/flows/sharedQueues.ts:12]: No, that's the parent queue's priority, this feature. (The children priorities are defined in the children array.) Now the two flow's jobs will have different priorities in these parent queues. Setting this property to 2 means the worker will check the prio1 queue first then the prio2 queue (not just the prio1). (The access flow's jobs will wait in the `access-check:1:waiting` queue, the status update jobs will wait in the `access-check:2:waiting` queue.)

Comment (TomiOhl) [reply src/flows/sharedQueues.ts:12]: Guess the reason why this was confusing to me is that it's in plural, thanks for the explanation

**Pull request merged**